File: bindings/java/examples/zebrad-t2z/src/main/java/com/zcash/t2z/examples/Keys.java
/**
 * Zcash Transparent Key Management
 *
 * Client-side key generation and signing for Zcash transparent addresses.
 * Compatible with Zebra (no wallet required).
 */
package com.zcash.t2z.examples;

import org.bitcoinj.core.Base58;
import org.bouncycastle.crypto.digests.RIPEMD160Digest;
import org.bouncycastle.crypto.params.ECPrivateKeyParameters;
import org.bouncycastle.crypto.signers.ECDSASigner;
import org.bouncycastle.jce.ECNamedCurveTable;
import org.bouncycastle.jce.spec.ECParameterSpec;
import org.bouncycastle.math.ec.ECPoint;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

public class Keys {
    // Zcash uses 2-byte version prefixes
    // Testnet/Regtest P2PKH: 0x1D25 -> 'tm' prefix
    private static final byte[] ZCASH_TESTNET_P2PKH = new byte[]{0x1d, 0x25};

    // WIF version byte for testnet
    private static final byte WIF_TESTNET = (byte) 0xef;

    // Pre-generated test keypair for regtest (deterministic for reproducibility)
    // WARNING: Only use this for testing! Never use in production.
    private static final String TEST_PRIVKEY_HEX = "e8f32e723decf4051aefac8e2c93c9c5b214313817cdb01a1494b917c8436b35";

    public static final ZcashKeypair TEST_KEYPAIR = keypairFromPrivateKey(hexToBytes(TEST_PRIVKEY_HEX));

    /**
     * Zcash keypair containing private key, public key, address, and WIF.
     */
    public static class ZcashKeypair {
        public final byte[] privateKey;
        public final byte[] publicKey;
        public final String address;
        public final String wif;

        public ZcashKeypair(byte[] privateKey, byte[] publicKey, String address, String wif) {
            this.privateKey = privateKey;
            this.publicKey = publicKey;
            this.address = address;
            this.wif = wif;
        }
    }

    /**
     * Create a keypair from an existing private key buffer.
     */
    public static ZcashKeypair keypairFromPrivateKey(byte[] privateKey) {
        if (privateKey.length != 32) {
            throw new IllegalArgumentException("Private key must be 32 bytes");
        }

        ECParameterSpec spec = ECNamedCurveTable.getParameterSpec("secp256k1");
        BigInteger privKeyInt = new BigInteger(1, privateKey);
        ECPoint pubKeyPoint = spec.getG().multiply(privKeyInt);

        // Get compressed public key (33 bytes)
        byte[] publicKey = pubKeyPoint.getEncoded(true);
        String address = pubkeyToAddress(publicKey);
        String wif = privateKeyToWIF(privateKey);

        return new ZcashKeypair(privateKey, publicKey, address, wif);
    }

    /**
     * Convert private key to WIF format.
     */
    public static String privateKeyToWIF(byte[] privateKey) {
        // WIF format: [version][privkey][compression flag][checksum]
        // Add compression flag (0x01 for compressed)
        byte[] payload = new byte[privateKey.length + 1];
        System.arraycopy(privateKey, 0, payload, 0, privateKey.length);
        payload[payload.length - 1] = 0x01; // compressed flag

        // Use the same encoding approach as address (manual Base58Check)
        return base58CheckEncode(new byte[]{WIF_TESTNET}, payload);
    }

    /**
     * Convert public key to Zcash transparent address.
     * Uses 2-byte version prefix for Zcash addresses.
     */
    public static String pubkeyToAddress(byte[] publicKey) {
        byte[] hash = hash160(publicKey);
        return base58CheckEncode(ZCASH_TESTNET_P2PKH, hash);
    }

    /**
     * Base58Check encode with arbitrary prefix.
     */
    public static String base58CheckEncode(byte[] version, byte[] payload) {
        byte[] data = new byte[version.length + payload.length + 4];
        System.arraycopy(version, 0, data, 0, version.length);
        System.arraycopy(payload, 0, data, version.length, payload.length);

        byte[] checksum = doubleSha256(Arrays.copyOfRange(data, 0, version.length + payload.length));
        System.arraycopy(checksum, 0, data, version.length + payload.length, 4);

        return Base58.encode(data);
    }

    /**
     * Hash160: SHA256 then RIPEMD160.
     */
    public static byte[] hash160(byte[] data) {
        byte[] sha256 = sha256(data);
        RIPEMD160Digest ripemd = new RIPEMD160Digest();
        ripemd.update(sha256, 0, sha256.length);
        byte[] result = new byte[20];
        ripemd.doFinal(result, 0);
        return result;
    }

    /**
     * SHA256 hash.
     */
    public static byte[] sha256(byte[] data) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return md.digest(data);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Double SHA256.
     */
    public static byte[] doubleSha256(byte[] data) {
        return sha256(sha256(data));
    }

    /**
     * Sign a message hash (sighash) with the keypair.
     * Returns 64-byte compact signature (r || s format).
     */
    public static byte[] signCompact(byte[] messageHash, ZcashKeypair keypair) {
        if (messageHash.length != 32) {
            throw new IllegalArgumentException("Message hash must be 32 bytes");
        }

        ECParameterSpec spec = ECNamedCurveTable.getParameterSpec("secp256k1");
        BigInteger privKeyInt = new BigInteger(1, keypair.privateKey);
        ECPrivateKeyParameters privKeyParams = new ECPrivateKeyParameters(
                privKeyInt,
                new org.bouncycastle.crypto.params.ECDomainParameters(
                        spec.getCurve(), spec.getG(), spec.getN(), spec.getH()
                )
        );

        ECDSASigner signer = new ECDSASigner();
        signer.init(true, privKeyParams);

        BigInteger[] signature = signer.generateSignature(messageHash);
        BigInteger r = signature[0];
        BigInteger s = signature[1];

        // Ensure low-S value (BIP 62)
        BigInteger halfN = spec.getN().shiftRight(1);
        if (s.compareTo(halfN) > 0) {
            s = spec.getN().subtract(s);
        }

        // Convert to 64-byte format (32 bytes r + 32 bytes s)
        byte[] result = new byte[64];
        byte[] rBytes = r.toByteArray();
        byte[] sBytes = s.toByteArray();

        // Handle potential leading zero in BigInteger representation
        int rStart = rBytes.length > 32 ? 1 : 0;
        int sStart = sBytes.length > 32 ? 1 : 0;
        int rLen = Math.min(32, rBytes.length - rStart);
        int sLen = Math.min(32, sBytes.length - sStart);

        System.arraycopy(rBytes, rStart, result, 32 - rLen, rLen);
        System.arraycopy(sBytes, sStart, result, 64 - sLen, sLen);

        return result;
    }

    /**
     * Convert bytes to hex string.
     */
    public static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    /**
     * Convert hex string to bytes.
     */
    public static byte[] hexToBytes(String hex) {
        if (hex.length() % 2 != 0) {
            throw new IllegalArgumentException("Hex string must have even length");
        }
        byte[] result = new byte[hex.length() / 2];
        for (int i = 0; i < result.length; i++) {
            result[i] = (byte) Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
        }
        return result;
    }
}

File: bindings/kotlin/examples/zebrad-regtest/src/main/kotlin/com/zcash/t2z/examples/Keys.kt
/**
 * Zcash Transparent Key Management
 *
 * Client-side key generation and signing for Zcash transparent addresses
 * Compatible with Zebra (no wallet required).
 */
package com.zcash.t2z.examples

import fr.acinq.bitcoin.Base58Check
import fr.acinq.bitcoin.Crypto
import fr.acinq.secp256k1.Secp256k1
import java.security.MessageDigest

// Zcash uses 2-byte version prefixes
// Testnet/Regtest P2PKH: 0x1D25 -> 'tm' prefix
// Mainnet P2PKH: 0x1CB8 -> 't1' prefix
private val ZCASH_TESTNET_P2PKH = byteArrayOf(0x1d, 0x25)
private val ZCASH_MAINNET_P2PKH = byteArrayOf(0x1c.toByte(), 0xb8.toByte())

// WIF version bytes
private const val WIF_TESTNET: Byte = 0xef.toByte()
private const val WIF_MAINNET: Byte = 0x80.toByte()

enum class NetworkType { MAINNET, TESTNET, REGTEST }

data class ZcashKeypair(
    val privateKey: ByteArray,
    val publicKey: ByteArray,
    val address: String,
    val wif: String
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ZcashKeypair) return false
        return privateKey.contentEquals(other.privateKey) &&
                publicKey.contentEquals(other.publicKey) &&
                address == other.address &&
                wif == other.wif
    }

    override fun hashCode(): Int {
        var result = privateKey.contentHashCode()
        result = 31 * result + publicKey.contentHashCode()
        result = 31 * result + address.hashCode()
        result = 31 * result + wif.hashCode()
        return result
    }
}

/**
 * Create a keypair from an existing private key buffer
 */
fun keypairFromPrivateKey(
    privateKey: ByteArray,
    network: NetworkType = NetworkType.REGTEST
): ZcashKeypair {
    require(privateKey.size == 32) { "Private key must be 32 bytes" }

    // pubkeyCreate returns 65-byte uncompressed, compress it to 33 bytes
    val uncompressedPubkey = Secp256k1.pubkeyCreate(privateKey)
    val publicKey = Secp256k1.pubKeyCompress(uncompressedPubkey)
    val address = pubkeyToAddress(publicKey, network)
    val wif = privateKeyToWIF(privateKey, network)

    return ZcashKeypair(privateKey, publicKey, address, wif)
}

/**
 * Convert private key to WIF format
 */
fun privateKeyToWIF(privateKey: ByteArray, network: NetworkType = NetworkType.REGTEST): String {
    val version = if (network == NetworkType.MAINNET) WIF_MAINNET else WIF_TESTNET
    // Add compression flag (0x01 for compressed) to the data
    val data = privateKey + byteArrayOf(0x01)
    return Base58Check.encode(version, data)
}

/**
 * Convert public key to Zcash transparent address
 * Uses 2-byte version prefix for Zcash addresses
 */
fun pubkeyToAddress(publicKey: ByteArray, network: NetworkType = NetworkType.REGTEST): String {
    // Hash160 the public key (SHA256 then RIPEMD160)
    val hash = hash160(publicKey)

    // Zcash uses 2-byte version prefix
    val versionBytes = if (network == NetworkType.MAINNET) ZCASH_MAINNET_P2PKH else ZCASH_TESTNET_P2PKH

    // Base58check encode with 2-byte prefix
    return Base58Check.encode(versionBytes, hash)
}

/**
 * Hash160: SHA256 then RIPEMD160
 */
fun hash160(data: ByteArray): ByteArray {
    val sha256 = MessageDigest.getInstance("SHA-256").digest(data)
    return Crypto.ripemd160(sha256)
}

/**
 * Double SHA256
 */
fun doubleSha256(data: ByteArray): ByteArray {
    val sha256 = MessageDigest.getInstance("SHA-256")
    return sha256.digest(sha256.digest(data))
}

/**
 * SHA256
 */
fun sha256(data: ByteArray): ByteArray {
    return MessageDigest.getInstance("SHA-256").digest(data)
}

/**
 * Sign a message hash (sighash) with the keypair
 * Returns 64-byte compact signature (r || s format)
 */
fun signCompact(messageHash: ByteArray, keypair: ZcashKeypair): ByteArray {
    require(messageHash.size == 32) { "Message hash must be 32 bytes" }
    val signature = Secp256k1.sign(messageHash, keypair.privateKey)
    // Return first 64 bytes (r || s), dropping the recovery byte
    return signature.copyOfRange(0, 64)
}

/**
 * Convert bytes to hex string
 */
fun bytesToHex(bytes: ByteArray): String = bytes.joinToString("") { "%02x".format(it) }

/**
 * Convert hex string to bytes
 */
fun hexToBytes(hex: String): ByteArray {
    require(hex.length % 2 == 0) { "Hex string must have even length" }
    return ByteArray(hex.length / 2) { i ->
        hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
    }
}

// Pre-generated test keypair for regtest (deterministic for reproducibility)
// WARNING: Only use this for testing! Never use in production.
private val TEST_PRIVKEY_HEX = "e8f32e723decf4051aefac8e2c93c9c5b214313817cdb01a1494b917c8436b35"

val TEST_KEYPAIR = keypairFromPrivateKey(hexToBytes(TEST_PRIVKEY_HEX), NetworkType.REGTEST)
